src/libcore/ptr.rs

```diff
@@ -2077,6 +2077,32 @@ impl<T: ?Sized> PartialEq for *mut T {
 #[stable(feature = "rust1", since = "1.0.0")]
 impl<T: ?Sized> Eq for *mut T {}
 
+#[unstable(feature = "ptr_dangling", issue = "45557")]
+/// Returns a const raw pointer that is dangling, but well-aligned and not null.
+///
+/// This is used where a non-null pointer is required.
+///
+/// # Safety
+///
+/// This function is safe to call, but the raw pointer it returns does not
+/// point to any value.
+pub const fn dangling<T>() -> *const T {
+    mem::align_of::<T>() as *const T
+}
+
+#[unstable(feature = "ptr_dangling", issue = "45557")]
+/// Returns a mut raw pointer that is dangling, but well-aligned and not null.
+///
+/// This is used where a non-null pointer is required.
+///
+/// # Safety
+///
+/// This function is safe to call, but the raw pointer it returns does not
+/// point to any value.
```

Comment:
nit: This is not necessarily true. For types with very large alignments, this may point to outside the 0th page (the page that’s guaranteed to be unmapped on most systems) and actually point to something (e.g. manually mapped 1st page). I would reword this to not mention value validity and rather mention that dereferencing the pointer is invalid, or something along the lines.

Reply:
My first draft was that dereferencing the pointer is invalid. While that's true in the "read memory" sense, in the higher level Rust semantic, dereferencing a pointer to ZST is fine.

Reply:
:smile: I'm having trouble finding the right words that navigate all of this.

Reply:
I think any `integer as *const T` pointer is invalid to use by these rules, since it's not based on anything http://llvm.org/docs/LangRef.html#pointer-aliasing-rules

So it must be treated as "truly dangling" or use some highly implementation defined way to access an actual value through that pointer.

Reply:
In that sense, it seems best to just say it does not point to any value. If it's equal to a bona fide pointer, that's incidental?

Reply:
> I think any integer_constant as *const T pointer is invalid to use by these rules

This isn’t exactly true, because pretty much any low level code that depends on stuff at well known addresses would be invalid (embedded/low level/OS code relies heavily on that, for example). I believe this rule from the page you’ve linked applies here:

> An integer constant other than zero or a pointer value returned from a function not defined within LLVM may be associated with address ranges allocated through mechanisms other than those provided by LLVM. Such ranges shall not overlap with any ranges of addresses allocated by mechanisms provided by LLVM.

This rule does not forbid creation and use of such address ranges, it only forbids overlap with address ranges generated by LLVM (alloca, malloc, ...).

Reply:
That makes sense. I have pushed a change that has no Safety section, because I have no useful advice apart from what's already in the other doc.